### PR TITLE
Auto mTLS for namespace/mesh beta peer authn policies

### PR DESCRIPTION
**Pull request by diemtvu**
This PR takes namespace and mesh `PeerAuthentication` into the consideration for configuring cluster socket.

This is an imperfect solution to make auto mTLS work with the new mTLS beta API, as it doesn't work correctly if workload-level or port-level policy is used.

Issue: https://github.com/istio/istio/issues/20746

### CODE REVIEW

**Comment by diemtvu** [comment]
@howardjohn @incfly This is to replace https://github.com/istio/istio/pull/20881, which need more time to investigate. I'll bring e2e test from the other while waiting on review. Thanks.

File: pilot/pkg/model/authentication.go [hunk 1]
```diff
@@ -59,6 +61,16 @@ type AuthenticationPolicies struct {
...
 
+	// namespaceMutualTLSMode is the MutualTLSMode correspoinding to the namespace-level PeerAuthentication.
+	// All namespace-level policies, and only them, are added to this map. If the policy mTLS mode is set
+	// to UNSET, it will be resolved to the globalMutualTLSMode, if exist (i.e not UNKNOWN), or MTLSPermissive
+	// otherwise.
+	namespaceMutualTLSMode map[string]MutualTLSMode
+
+	// globalMutualTLSMode is the MutualTLSMode corresponding to the mesh-level PeerAuthentication.
+	// This value can be MTLSUnknown, if there is no mesh-level policy.
+	globalMutualTLSMode MutualTLSMode
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: you don't have to duplicate this mode. we can always calculate this value on demand. by policy[rootns], no?

Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Just more convenient and readable to use a dedicated variable, imo. It needs to be treated slightly different, with all the fallback and inheritance logics.

Comment (incfly) [reply pilot/pkg/model/authentication.go]: similarly to `namespaceMutualTLSMode`.

File: pilot/pkg/model/authentication.go [hunk 2]
```diff
@@ -97,13 +110,75 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
 	}
 }
 
+func apiModeToMutualTLSMode(mode v1beta1.PeerAuthentication_MutualTLS_Mode) MutualTLSMode {
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: this is duplication with policy applier's `getMutualTlsmode` right?

File: pilot/pkg/model/push_context_test.go [hunk 3]
```diff
@@ -649,6 +651,168 @@ func TestSidecarScope(t *testing.T) {
 	}
 }
 
+func TestBestEffortInferServiceMTLSMode(t *testing.T) {
+	const testNamespace string = "test-namespace"
+	ps := NewPushContext()
+	env := &Environment{Watcher: mesh.NewFixedWatcher(&meshconfig.MeshConfig{RootNamespace: "istio-system"})}
+	ps.Mesh = env.Mesh()
+	ps.ServiceDiscovery = env
+	authNPolicies := map[string]*authn.Policy{
```
Comment (incfly) [inline pilot/pkg/model/push_context_test.go]: might add an external one for UNKONWN.

File: pilot/pkg/model/push_context_test.go [hunk 4]
```diff
@@ -649,6 +651,168 @@ func TestSidecarScope(t *testing.T) {
 	}
 }
 
+func TestBestEffortInferServiceMTLSMode(t *testing.T) {
+	const testNamespace string = "test-namespace"
```
Comment (incfly) [inline pilot/pkg/model/push_context_test.go]: make `beta-namespace` and `new-namespace` as constant var as well to be consistent

Comment (incfly) [reply pilot/pkg/model/push_context_test.go]: clearer they're different ns.

File: pilot/pkg/model/authentication.go [hunk 5]
```diff
@@ -97,13 +110,75 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+
 func (policy *AuthenticationPolicies) addPeerAuthentication(configs []Config) {
+	foundNamespaceMTLS := make(map[string]v1beta1.PeerAuthentication_MutualTLS_Mode)
+
 	for _, config := range configs {
 		policy.peerAuthentications[config.Namespace] =
 			append(policy.peerAuthentications[config.Namespace], config)
+
+		// Mesh & namespace level policy must be named as `default` (config_consts.DefaultAuthenticationPolicyName)
+		// per validation rule. Extra check for empty selector is optional.
```
Comment (incfly) [inline pilot/pkg/model/authentication.go]: are you saying the the name equals to `default` is enough? and we don't have to check selector?

Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: With validation on, it should be equivalent. But thinking a bit more on this, may be we should check both just to be safe.

File: pilot/pkg/model/push_context_test.go (hunk 4 above)
Comment (diemtvu) [reply pilot/pkg/model/push_context_test.go]: done

File: pilot/pkg/model/push_context_test.go (hunk 3 above)
Comment (diemtvu) [reply pilot/pkg/model/push_context_test.go]: done

File: pilot/pkg/model/authentication.go (hunk 2 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: yeah :). Need some refactoring, may be when we deprecate alpha.

File: pilot/pkg/model/authentication.go (hunk 1 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: namespace is even more work, as it need find the one name default (and/or have no selector). It is more or less the same as my attempt before which use the applier (much simpler code, but then slightly less efficient).

File: pilot/pkg/model/authentication.go [hunk 6]
```diff
@@ -97,13 +110,75 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+}
+
 func (policy *AuthenticationPolicies) addPeerAuthentication(configs []Config) {
+	foundNamespaceMTLS := make(map[string]v1beta1.PeerAuthentication_MutualTLS_Mode)
+
 	for _, config := range configs {
 		policy.peerAuthentications[config.Namespace] =
 			append(policy.peerAuthentications[config.Namespace], config)
+
+		// Mesh & namespace level policy must be named as `default` (config_consts.DefaultAuthenticationPolicyName)
```
Comment (howardjohn) [inline pilot/pkg/model/authentication.go]: Can we revert this change? This will break multiple control plane.. why do we need an arbitrary restriction?

File: pilot/pkg/model/authentication.go (hunk 5 above)
Comment (incfly) [reply pilot/pkg/model/authentication.go]: yes. it's building on some convention, (you could say that's part of API), but checking selector seems more robust.

File: pilot/pkg/model/authentication.go (hunk 6 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: We want to enforce singleton. Why does it break multiple control plane btw?

Comment (incfly) [reply pilot/pkg/model/authentication.go]: i think it's already there, this limitation.

with multi control plane, will they be lived in the same ns? if they are different, we can have two `rootNS`, and each is named as default. will that work?

Comment (howardjohn) [reply pilot/pkg/model/authentication.go]: Because its no longer a singleton with multiple control planes. I should be able to  define one policy for ns `foo` with `istio.io/rev=istiod-1` and another in ns `foo` with `istio.io/rev=istiod=2`. Then it is still a "singleton" from the perspective of each namespace.

I suppose it does *break* this, it just means you cannot do canarys of this config

File: pilot/pkg/model/authentication.go (hunk 1 above)
Comment (incfly) [reply pilot/pkg/model/authentication.go]: its still possible with cleaner and efficiency. if you store namespace policy separately in the `model/authentication.go`, not mingled with ones having selector in the same `peer` field. and then in the NewApplier, passing namespace configs precisely, plus an empty label selector. empty jwt policy, etc.

i think it's fine to leave as it is as release cut approaching, above is just a way to hold on to package separate

File: pilot/pkg/model/authentication.go (hunk 6 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Hmm, ok I change this to check selector. Need to send out different PR to drop validation if that the case, though I don't see how can we enforce the singleton then

File: pilot/pkg/model/authentication.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/authentication.go]: Actually, they are same

File: pilot/pkg/model/authentication.go [hunk 7]
```diff
@@ -97,13 +109,74 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+}
+
 func (policy *AuthenticationPolicies) addPeerAuthentication(configs []Config) {
+	foundNamespaceMTLS := make(map[string]v1beta1.PeerAuthentication_MutualTLS_Mode)
+
 	for _, config := range configs {
 		policy.peerAuthentications[config.Namespace] =
 			append(policy.peerAuthentications[config.Namespace], config)
+
+		// Mesh & namespace level policy are those that have empty selector.
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/authentication.go]: Do we also have the name=`default` limit?

File: pilot/pkg/model/authentication.go [hunk 8]
```diff
@@ -97,13 +109,74 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+		spec := config.Spec.(*v1beta1.PeerAuthentication)
+		if spec.Selector == nil || len(spec.Selector.MatchLabels) == 0 {
+			mode := v1beta1.PeerAuthentication_MutualTLS_UNSET
+			if spec.Mtls != nil {
+				mode = spec.Mtls.Mode
+			}
+			if config.Namespace == policy.rootNamespace {
+				// This is mesh-level policy. Resolve the UNSET mode if needed, and update the foundGlobalMTLS.
+				if mode == v1beta1.PeerAuthentication_MutualTLS_UNSET {
+					policy.globalMutualTLSMode = MTLSPermissive
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/authentication.go]: Can you add comment why this?

File: pilot/pkg/model/authentication.go [hunk 9]
```diff
@@ -97,13 +109,74 @@ func (policy *AuthenticationPolicies) addRequestAuthentication(configs []Config)
...
+				} else {
+					policy.globalMutualTLSMode = apiModeToMutualTLSMode(mode)
+				}
+			} else {
+				// For regular namespace, just add to the intemediate map.
+				foundNamespaceMTLS[config.Namespace] = mode
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/authentication.go]: The latter one will override.

File: pilot/pkg/model/authentication.go (hunk 8 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/authentication.go]: Also this duplicate with L159

File: pilot/pkg/model/push_context.go [hunk 10]
```diff
@@ -1812,3 +1812,33 @@ func (ps *PushContext) NetworkGatewaysByNetwork(network string) []*Gateway {
...
+// BestEffortInferServiceMTLSMode infers the mTLS mode for the service + port from all authentication
+// policies (both alpha and beta) in the system. The function always returns MTLSUnknown for external service.
+// The resulst is a best effort. It is because the PeerAuthentication is workload-based, this function is unable
+// to compute the correct service mTLS mode without knowing service to workload binding. For now, this
+// function uses only mesh and namespace level PeerAuthentication and ignore workload & port level policies.
+// This function is used to give a hint for auto-mTLS configuration on client side.
+func (ps *PushContext) BestEffortInferServiceMTLSMode(service *Service, port *Port) MutualTLSMode {
+	if service.MeshExternal {
+		// Only need the authentication MTLS mode when service is not external.
+		return MTLSUnknown
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go]: should be MTLSDisable

**Comment by hzxuzhonghu** [comment]
> This is an imperfect solution to make auto mTLS work with the new mTLS beta API, as it doesn't work correctly if workload-level or port-level policy is used.

How do you plan to handle this?

File: pilot/pkg/model/authentication.go (hunk 7 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: No, we will remove that.

File: pilot/pkg/model/authentication.go (hunk 8 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Done

File: pilot/pkg/model/authentication.go (hunk 2 above)
Comment (incfly) [reply pilot/pkg/model/authentication.go]: ok. leave a todo on the top of func then.

Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Done

File: pilot/pkg/model/authentication.go [hunk 11]
```diff
@@ -59,6 +60,16 @@ type AuthenticationPolicies struct {
 
 	peerAuthentications map[string][]Config
 
+	// namespaceMutualTLSMode is the MutualTLSMode correspoinding to the namespace-level PeerAuthentication.
+	// All namespace-level policies, and only them, are added to this map. If the policy mTLS mode is set
+	// to UNSET, it will be resolved to the globalMutualTLSMode, if exist (i.e not UNKNOWN), or MTLSPermissive
```
[the remark this answers is not in the document]
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: Rephrase.

File: pilot/pkg/model/authentication.go (hunk 9 above)
Comment (diemtvu) [reply pilot/pkg/model/authentication.go]: We expect to have at most one namespace-level per namespace. Same as mesh level.

Anyway, I add a bookkeeping to make sure only the oldest one is used.

File: pilot/pkg/model/push_context.go (hunk 10 above)
Comment (diemtvu) [reply pilot/pkg/model/push_context.go]: This is to keep the same behavior as before. Maybe keep it as is to avoid more regression.

**Comment by diemtvu** [comment]
> > This is an imperfect solution to make auto mTLS work with the new mTLS beta API, as it doesn't work correctly if workload-level or port-level policy is used.
> 
> How do you plan to handle this?

We still investigate other approaches (e.g https://github.com/istio/istio/pull/20881). For now we have to ship with this limitation.

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 12]
```diff
@@ -210,12 +209,7 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(proxy *model.Proxy,
 			}
 
 			setUpstreamProtocol(proxy, defaultCluster, port, model.TrafficDirectionOutbound)
-			serviceMTLSMode := model.MTLSUnknown
-			if !service.MeshExternal {
-				// Only need the authentication MTLS mode when service is not external.
-				policy, _ := push.AuthenticationPolicyForWorkload(service, port)
-				serviceMTLSMode = authn_v1alpha1_applier.GetMutualTLSMode(policy)
-			}
+			serviceMTLSMode := push.BestEffortInferServiceMTLSMode(service, port)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster.go]: I don't really understand this. Why are we replacing something that is correct with something that is "best effort"? Am I missing something here?

Comment (diemtvu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go]: With the workload-oriented policy, we cannot tell what is the mTLS mode of the`service` anymore. 
It's still correct if beta policy is not used and fallback to alpha.

**Comment by diemtvu** [comment]
+ @louiscryan for TOC review

FYI, this PR partially implement auto-mTLS for the new beta API. Due to workload selector, we are no longer able to know the mTLS mode for a `service` , so the feature is limited to namespace & mesh policies.

In term of performance, it adds an extra map look up on pilot for each cluster. The map, keyed by namespace, is created once during push context init. Overall, the overhead cpu/memory is negligible.

E2E tests are included in this PR, verifying the new behavior, as well as the existing behavior when beta API is not used, or when there are a mix of alpha and beta API.

**Comment by rshriram** [comment]
@howardjohn can you summarize your objection here..

**Comment by howardjohn** [comment]
I don't have specific objections to this PR (unlike the previous PR doing the same thing - this one looks much safer), I haven't had a chance to give it a thorough review yet. But I dont see why we are backporting new features given we had code freeze weeks ago

**Comment by louiscryan** [comment]
@diemtvu just so I'm understanding. With the workload oriented policy we can't infer the mTLS setting to apply to the Envoy cluster unless

@howardjohn don't we already have code the knows the set of endpoints within a service. Can't we maintain a running rollup of the mTLS policy applied to the workloads within each cluster definition ?

@rshriram should we collectively conference on this tomorrow?

**Comment by incfly** [comment]
> don't we already have code the knows the set of endpoints within a service. Can't we maintain a running rollup of the mTLS policy applied to the workloads within each cluster definition ?

https://github.com/istio/istio/pull/20881 is doing something like that. we stopped because endpoint finding only updating endpoint label without updating cluster config somehow have some sticky effect making old transport socket config retained for the same host. there might be some workaround.

**Comment by diemtvu** [comment]
> @diemtvu just so I'm understanding. With the workload oriented policy we can't infer the mTLS setting to apply to the Envoy cluster unless
> 
> @howardjohn don't we already have code the knows the set of endpoints within a service. Can't we maintain a running rollup of the mTLS policy applied to the workloads within each cluster definition ?
> 
> @rshriram should we collectively conference on this tomorrow?

That's right.

**Comment by rshriram** [comment]
So the problem is that ifI define a workload specific authN policy, then for each endpoint in every cluster, we have to run through all these workload specific policies to identify the workload specific policy if any and then set the tls context just for that endpoint. This reverse lookup is costly IMO, for every pod. The previous PR managed to do this but the current PR does not. Ergo, with the current PR, if the user has a workload specific policy to disable mTLS, then she would also have to declare a destination rule targeting that workload/subset disabling mTLS.

**Comment by louiscryan** [comment]
@rshriram I agree the reverse join is likely not practical and would be removed in 1.6 anyway. 

I'm fine to proceed with the PR as the current state in the release branch is broken.

It's clear that the use of selectors in the Beta API introduces complexity for the end-user that needs to be thoroughly accounted for in documentation. In general however that complexity largely mirrors the complexity with the Alpha API for similarly 'selective' policies. The Beta API still has a clear benefit in terms of policy determinism.

@howardjohn will complete a more thorough review before this should be committed but I am approving this now.

File: pilot/pkg/networking/core/v1alpha3/cluster_test.go [hunk 13]
```diff
@@ -377,6 +384,22 @@ func buildTestClustersWithProxyMetadataWithIps(serviceHostname string, serviceRe
...
 					}}, nil
 			}
+			if typ == collections.IstioSecurityV1Beta1Peerauthentications.Resource().GroupVersionKind() && peerAuthn != nil {
+				policyName := "default"
+				if peerAuthn.Selector != nil {
+					policyName = "acme"
+				}
+				return []model.Config{
+					{ConfigMeta: model.ConfigMeta{
+						Type:      collections.IstioSecurityV1Beta1Peerauthentications.Resource().Kind(),
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster_test.go]: nit: add group here as well
